libbeat/processors/add_cloud_metadata/provider_aws_ec2_test.go

```diff
@@ -75,23 +76,159 @@ func TestRetrieveAWSMetadata(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	actual, err := p.Run(&beat.Event{Fields: common.MapStr{}})
+	cases := []struct {
+		fields          common.MapStr
+		expectedResults common.MapStr
+	}{
+		{
+			common.MapStr{},
+			common.MapStr{
+				"cloud": common.MapStr{
+					"provider": "ec2",
+					"instance": common.MapStr{
+						"id": "i-11111111",
+					},
+					"machine": common.MapStr{
+						"type": "t2.medium",
+					},
+					"region":            "us-east-1",
+					"availability_zone": "us-east-1c",
+				},
+			},
+		},
+		{
+			common.MapStr{
+				"cloud": common.MapStr{
+					"instance": common.MapStr{
+						"id": "i-000",
+					},
+				},
+			},
+			common.MapStr{
+				"cloud": common.MapStr{
+					"instance": common.MapStr{
+						"id": "i-000",
+					},
+				},
+			},
+		},
+		{
+			common.MapStr{
+				"cloud.provider": "ec2",
+			},
+			common.MapStr{
+				"cloud.provider": "ec2",
+				"cloud": common.MapStr{
+					"provider": "ec2",
+					"instance": common.MapStr{
+						"id": "i-11111111",
+					},
+					"machine": common.MapStr{
+						"type": "t2.medium",
+					},
+					"region":            "us-east-1",
+					"availability_zone": "us-east-1c",
+				},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		actual, err := p.Run(&beat.Event{Fields: c.fields})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.Equal(t, c.expectedResults, actual.Fields)
+	}
+}
+
+func TestRetrieveAWSMetadataOverwriteTrue(t *testing.T) {
+	logp.TestingSetup()
+
+	server := initEC2TestServer()
+	defer server.Close()
+
+	config, err := common.NewConfigFrom(map[string]interface{}{
+		"host":      server.Listener.Addr().String(),
+		"overwrite": true,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err := New(config)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	expected := common.MapStr{
-		"cloud": common.MapStr{
-			"provider": "ec2",
-			"instance": common.MapStr{
-				"id": "i-11111111",
+	cases := []struct {
+		fields          common.MapStr
+		expectedResults common.MapStr
+	}{
+		{
+			common.MapStr{},
+			common.MapStr{
+				"cloud": common.MapStr{
+					"provider": "ec2",
+					"instance": common.MapStr{
+						"id": "i-11111111",
+					},
+					"machine": common.MapStr{
+						"type": "t2.medium",
+					},
+					"region":            "us-east-1",
+					"availability_zone": "us-east-1c",
+				},
 			},
-			"machine": common.MapStr{
-				"type": "t2.medium",
+		},
+		{
+			common.MapStr{
+				"cloud": common.MapStr{
+					"instance": common.MapStr{
+						"id": "i-000",
+					},
+				},
+			},
+			common.MapStr{
+				"cloud": common.MapStr{
+					"provider": "ec2",
+					"instance": common.MapStr{
+						"id": "i-11111111",
+					},
+					"machine": common.MapStr{
+						"type": "t2.medium",
+					},
+					"region":            "us-east-1",
+					"availability_zone": "us-east-1c",
+				},
+			},
+		},
+		{
+			common.MapStr{
+				"cloud.provider": "ec2",
+			},
+			common.MapStr{
+				"cloud.provider": "ec2",
```

Comment:
Is that really what we expect? I would say this is a bug?

Reply:
I do think it's a bug. When `overwrite = false`, if field key is `cloud.provider`, the code should recognize the `cloud` part and not overwrite it.

Reply:
@ruflin I added some code to manually check the field key to see if `cloud` is a part or not https://github.com/elastic/beats/pull/11612/files#diff-d838333d9237e9c88a650a5d20f09a73R372

Reply:
@ruflin Do you think `add_cloud_metadata` processor is the place to deal with dot in field key?